lib/Differentiator/ReverseModeVisitor.cpp

```diff
@@ -4100,31 +4124,150 @@ Expr* getArraySizeExpr(const ArrayType* AT, ASTContext& context,
     return {nullptr, nullptr};
   }
 
-  // FIXME: Add support for differentiating calls to constructors.
-  // We currently assume that constructor arguments are non-differentiable.
   StmtDiff
   ReverseModeVisitor::VisitCXXConstructExpr(const CXXConstructExpr* CE) {
-    llvm::SmallVector<Expr*, 4> clonedArgs;
+
+    llvm::SmallVector<Expr*, 4> primalArgs;
+    llvm::SmallVector<Expr*, 4> adjointArgs;
+    llvm::SmallVector<Expr*, 4> reverseForwAdjointArgs;
+    // It is used to store '_r0' temporary gradient variables that are used for
+    // differentiating non-reference args.
+    llvm::SmallVector<Stmt*, 4> prePullbackCallStmts;
+
+    // Insertion point is required because we need to insert pullback call
+    // before the statements inserted by 'Visit(arg, ...)' calls for arguments.
+    std::size_t insertionPoint = getCurrentBlock(direction::reverse).size();
+
+    // FIXME: Restore arguments passed as non-const reference.
     for (const auto* arg : CE->arguments()) {
-      auto argDiff = Visit(arg, dfdx());
-      clonedArgs.push_back(argDiff.getExpr());
+      StmtDiff argDiff{};
+      Expr* adjointArg = nullptr;
+      if (utils::IsReferenceOrPointerArg(arg)) {
+        argDiff = Visit(arg);
+        adjointArg = argDiff.getExpr_dx();
+      } else {
+        // non-reference arguments are differentiated as follows:
+        //
+        // primal code:
+        // ```
+        // SomeClass c(u, ...);
+        // ```
+        //
+        // Derivative code:
+        // ```
+        // // forward pass
+        // ...
+        // // reverse pass
+        // double _r0 = 0;
+        // SomeClass_pullback(c, u, ..., &_d_c, &_r0, ...);
+        // _d_u += _r0;
+        QualType dArgTy = getNonConstType(arg->getType(), m_Context, m_Sema);
+        VarDecl* dArgDecl = BuildVarDecl(dArgTy, "_r", getZeroInit(dArgTy));
+        prePullbackCallStmts.push_back(BuildDeclStmt(dArgDecl));
+        adjointArg = BuildDeclRef(dArgDecl);
+        argDiff = Visit(arg, BuildDeclRef(dArgDecl));
+      }
+
+      if (utils::isArrayOrPointerType(arg->getType())) {
+        reverseForwAdjointArgs.push_back(adjointArg);
+        adjointArgs.push_back(adjointArg);
+      } else {
+        if (utils::isReference(arg))
```

Comment:
```suggestion
        if (arg->IgnoreParenImpCasts()->getType()->isReferenceType())
```
I wonder if that won't just work?

Reply:
we still need to get the decl, because the expression itself can't tell if it's a reference or not for some reason, so I guess defining a function is more general. plus, this function can then be reused, as this seems to be a common task